docs/assets.rst

```diff
@@ -0,0 +1,673 @@
+change a file, refresh the browser, and see the effects of the changed file).
+For production, the ``staticfiles`` app provides the management command
+``collectstatic``, which gathers files to the ``/static`` folder for efficent
+file serving.
+
+The Django documents for ``staticfiles`` are mostly focused on usage.
+Additional details are needed to understand how ``django-pipeline``
+customizes ``staticfiles``.
+
+Configuration
+*************
+
+The ``staticfiles`` app is configured by Django settings:
+
+``STATIC_ROOT``
+   The folder on the file system where assets are collected. For MDN, this is
+   the ``static`` folder in the ``kuma`` directory.
+``STATIC_URL``
+   The base URL for static assets. In development, this is
+   ``http://localhost:8000/static/``, and in production it is
+   ``https://developer.mozilla.org/static/``.
+``STATICFILES_FINDERS``
+   The dotted path to classes implementing ``staticfiles`` :ref:`Finder <staticfiles-finder>`.
+   These determine what files will be collected and served. Kuma uses four finders:
+
+   * `django.contrib.staticfiles.finders.FileSystemFinder`_: Finds files
+     in folders specified by ``STATICFILES_DIRS``
+   * `django.contrib.staticfiles.finders.AppDirectoriesFinder`_: Finds
+     files in the ``static`` subfolder of any installed apps
+   * `pipeline.finders.CachedFileFinder`_: Strips hashes from filenames to
+     identify the "pre-cached" names for files.
+   * `pipeline.finders.PipelineFinder`_: When combined assets are not enabled
+     (``PIPELINE['PIPELINE_ENABLED'] == False``), returns the source
+     files instead of the combined bundle file.
+
+``STATICFILES_DIRS``
+   A list of folders in the ``kuma`` directory that the ``FileSystemFinder``
+   will scan for static assets. For MDN, this is set to ``kuma/static``,
+   ``build/locale``, and ``jinja2/includes/icons``.
+
+   For example, the :ref:`localization JavaScript <build-locale-js>`
+   ``build/locale/jsi18n/fr/javascript.js`` will be collected to
+   ``static/jsi18n/fr/javascript.js``.
+``STATICFILES_STORAGE``
+   The dotted path to a class implementing ``staticfiles``
+   :ref:`Storage <staticfiles-storage>`.  Storage determines where files
+   are stored, what URLs they have, and provides hooks for modifying files when
+   copying them.  Kuma uses three different storages, depending on the context:
+
+   * Development server (``DEBUG=False``): pipeline.storage.NonPackagingPipelineStorage_,
+     which avoids combining files when collecting them.
+   * Production server (``DEBUG=True``): kuma.core.pipeline.storage.ManifestPipelineStorage_,
+     which combines packaged files, hashes the names, and creates a manifest.
+   * Testing (``pytest``, TravisCI, etc.) and ``make`` commands: pipeline.storage.PipelineStorage_,
+     which combines packaged files but does not hash the names.
+
+.. _django.contrib.staticfiles: cs.djangoproject.com/en/1.11/ref/contrib/staticfiles/
+
+.. _staticfiles-finder:
+
+Finder classes
```

Comment:
I don't have any objections to the sections here and below on Finder and Storage classes, but I'm not sure why I need to understand things at this level of detail. Maybe I'm missing something. Is there a configuration setting somehwere that controls which finders and storage classes are used in which circumstances? Perhaps I'd understand this better if it was clear which parts are under our control and which are just the things that staticfiles and django-pipeline do for us by default.

Reply:
Some of this is just "I had to understand this, so now you need to as well".

Our configuration points are what finders we use (and what order), which storage we use (just one at a time), and how we configure them. We also need to treat this as something we control, rather than something that we work around.

A lot of this is necessary because ``django-pipeline`` uses the framework to do "magic", and you have to understand the framework to know why we have at least 3 different configurations. I hope to get rid of ``django-pipeline``, and then a lot of this may become irrelevant.